src/source/Ice/IceAgent.c

```diff
@@ -1444,8 +1484,14 @@ STATUS iceAgentReadyStateSetup(PIceAgent pIceAgent)
     // no state timeout for ready state
     pIceAgent->stateEndTime = INVALID_TIMESTAMP_VALUE;
 
+    relayCandidateSelected = IS_CANN_PAIR_SENDING_FROM_RELAYED(pIceAgent->pDataSendingIceCandidatePair);
+
+    // unlock ice lock before calling turn api.
+    MUTEX_UNLOCK(pIceAgent->lock);
```

Comment:
This is a race condition case and a layer violation. Also, you would rather unlock inside the if statement. Is there any other simple way to handle this case in the outer functionality outside of the lock?

It might not be neat but think if the timer queue can be scheduled for this

Reply:
removed the locking. IceAgent can now call turnConnection while holding the iceAgent lock because I made turnConnection to release its lock before triggering ice api. So there is no more circular locking.